dynamicscrm/metadata.go

```diff
@@ -34,109 +34,181 @@ func (c *Connector) listObjectMetadata(
 	}
 
-	result, err := convertEntitySetToMetadataSet(objectNames, entities)
+	// Our goal is to get properties that can be found in Read response.
+	// Not all Attributes are real fields.
+	// IsValidODataAttribute == true -> filters fields returned in the Read payload.
+	// Now we have all good fields that certainly will be in payload.
+	// But if they are returned doesn't mean you can request them!
+	// IsValidForRead == true -> filters fields that can be requested in GET query params.
+	url.WithQueryParam("$filter", "(IsValidODataAttribute eq true and IsValidForRead eq true)")
+	// We cannot use $select clause to scope response, unfortunately, `Targets` field breaks $select.
+	// Falling back to requesting the whole payload.
+
+	body, err := c.performGetRequest(ctx, url)
 	if err != nil {
 		return nil, err
 	}
 
-	return &common.ListObjectMetadataResult{
-		Result: result,
-		Errors: nil,
-	}, nil
+	fields, err := extractFields(body)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", err, objectName)
+	}
+
+	return fields, nil
 }
 
-// collects field properties and groups them in entities, other data in XML is ignored.
-func extractEntities(root *xquery.XML) (EntitySet, error) {
-	querySchema := fmt.Sprintf("//edmx:DataServices/Schema[@Namespace='%v']", CRMMetadataSchemaName)
-
-	schema := root.FindOne(querySchema)
-	if schema.IsEmpty() {
-		return nil, ErrMissingSchema
-	}
-
-	entities := NewEntitySet()
-	// List all field properties that exist for current schema
-	queryListAllSchemaProperties := fmt.Sprintf(
-		"//edmx:DataServices/Schema[@Namespace='%v']/EntityType[*]/Property/@Name", CRMMetadataSchemaName)
-	for _, nameAttr := range root.FindMany(queryListAllSchemaProperties) {
-		// parent of a property is an Entity.
-		// Entity may inherit properties from a parent
-		// We save entity name and the name of its parent, so later we can infer all properties by denormalisation
-		property := nameAttr.Parent()
-		entityObj := property.Parent()
-		entityName := entityObj.Attr("Name")
-		parentName := entityObj.Attr("BaseType")
-		entity := entities.GetOrCreate(entityName, parentName)
-		propertyName := property.Attr("Name")
-		entity.AddProperty(propertyName)
-	}
-
-	queryListAbstractEntities := fmt.Sprintf(
-		"//edmx:DataServices/Schema[@Namespace='%v']/EntityType[@Abstract='true']", CRMMetadataSchemaName)
-	for _, abstractEntity := range root.FindMany(queryListAbstractEntities) {
-		if !abstractEntity.HasChildren() {
-			// these entities were not included by previous query as they have no properties
-			// we programmatically find these special types, which are "primary values" but for structs
-			// Ex: crmbaseentity, crmmodelbaseentity,
-			entityName := abstractEntity.Attr("Name")
-			// effectively only create
-			_ = entities.GetOrCreate(entityName, "")
+// Given Collection of attributes from endpoint response, select those applicable to Read response.
+func extractFields(attributes *ajson.Node) (map[string]string, error) {
```

Comment:
`selectValidReadFields` or something similar would read better

Reply:
Changed the comment it was a bit misleading. Function is `extractFieldsFromJSON`. The input is parsed, there is no selection going on. API request covered the filtering part.